File: trunk/src/main/java/net/esper/core/EPPatternStmtStartMethod.java
package net.esper.core;

import net.esper.pattern.PatternStarter;
import net.esper.pattern.PatternStopCallback;
import net.esper.pattern.PatternContext;
import net.esper.pattern.PatternMatchCallback;
import net.esper.schedule.ScheduleBucket;

/**
 * Method for starting a pattern statement.
 */
public class EPPatternStmtStartMethod
{
    private final EPServicesContext services;
    private final PatternStarter patternStarter;
    private final ScheduleBucket scheduleBucket;

    /**
     * Ctor.
     * @param services - services for starting
     * @param patternStarter - pattern start handle
     */
    public EPPatternStmtStartMethod(EPServicesContext services, PatternStarter patternStarter)
    {
        this.services = services;
        this.patternStarter = patternStarter;

        // Allocate the statement's schedule bucket which stays constant over it's lifetime.
        // The bucket allows callbacks for the same time to be ordered (within and across statements) and thus deterministic.
        scheduleBucket = services.getSchedulingService().allocateBucket();
    }

    /**
     * Start pattern.
     * @param matchCallback - callback by pattern when matches are found
     * @return stop method used for stopping pattern
     */
    public EPStatementStopMethod start(PatternMatchCallback matchCallback)
    {
        PatternContext context = new PatternContext(services.getFilterService(), services.getSchedulingService(), scheduleBucket);
        final PatternStopCallback stopCallback = patternStarter.start(matchCallback, context);

        EPStatementStopMethod stopMethod = new EPStatementStopMethod()
        {
            public void stop()
            {
                stopCallback.stop();
            }
        };

        return stopMethod;
    }

}

File: NEsper/core/EPPatternStmtStartMethod.cs
using System;

using net.esper.pattern;
using net.esper.schedule;
using net.esper.util;

namespace net.esper.core
{
	/// <summary>
    /// Method for Starting a pattern statement.
    /// </summary>
	
    public class EPPatternStmtStartMethod
	{
		private readonly EPServicesContext services;
		private readonly PatternStarter patternStarter;
		private readonly ScheduleBucket scheduleBucket;
		
		/// <summary> Ctor.</summary>
		/// <param name="services">services for Starting
		/// </param>
		/// <param name="patternStarter">pattern Start handle
		/// </param>
		
        public EPPatternStmtStartMethod(EPServicesContext services, PatternStarter patternStarter)
		{
			this.services = services;
			this.patternStarter = patternStarter;
			
			// Allocate the statement's schedule bucket which stays constant over it's lifetime.
			// The bucket allows callbacks for the same time to be ordered (within and across statements) and thus deterministic.
			scheduleBucket = services.SchedulingService.allocateBucket();
		}
		
		/// <summary> Start pattern.</summary>
		/// <param name="matchCallback">callback by pattern when matches are found
		/// </param>
		/// <returns> Stop method used for Stopping pattern
		/// </returns>

        public virtual EPStatementStopMethod Start(PatternMatchCallback matchCallback)
		{
			PatternContext context = new PatternContext(services.FilterService, services.SchedulingService, scheduleBucket, services.EventAdapterService);
			StopCallback stopCallback = patternStarter.Start(matchCallback, context);
			EPStatementStopMethod stopMethod = new EPStatementStopMethod(
				delegate() { stopCallback(); } );
			
			return stopMethod;
		}
	}
}
